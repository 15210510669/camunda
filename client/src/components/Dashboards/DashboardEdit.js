/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import moment from 'moment';
import update from 'immutability-helper';
import {evaluateReport} from 'services';

import {DashboardRenderer, EntityNameForm} from 'components';

import {AddButton} from './AddButton';
import {Grid} from './Grid';
import {DimensionSetter} from './DimensionSetter';
import {DeleteButton} from './DeleteButton';
import {DragBehavior} from './DragBehavior';
import {ResizeHandle} from './ResizeHandle';

export default class DashboardEdit extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      reports: props.initialReports,
      addButtonVisible: true
    };
  }

  updateReport = ({report, ...changes}) => {
    const reportIdx = this.state.reports.indexOf(report);

    Object.keys(changes).forEach(prop => {
      changes[prop] = {$set: changes[prop]};
    });

    this.setState({
      reports: update(this.state.reports, {
        [reportIdx]: changes
      })
    });
  };

  showAddButton = () => {
    this.setState({
      addButtonVisible: true
    });
  };

  hideAddButton = () => {
    this.setState({
      addButtonVisible: false
    });
  };

  addReport = newReport => {
    this.setState({reports: update(this.state.reports, {$push: [newReport]})});
  };

  deleteReport = ({report: reportToRemove}) => {
    this.setState({
      reports: this.state.reports.filter(report => report !== reportToRemove)
    });
  };

  render() {
    const {name, lastModifier, lastModified, id, isNew} = this.props;

    const {reports} = this.state;

    return (
      <div className="DashboardEdit">
        <div className="header">
          <EntityNameForm
            id={id}
            initialName={name}
            lastModified={lastModified}
            lastModifier={lastModifier}
            autofocus={isNew}
            entity="Dashboard"
            onSave={(evt, newName) => this.props.saveChanges(newName, reports)}
          />
          <div className="subHead">
            <div className="metadata">
              Last modified {moment(lastModified).format('lll')} by {lastModifier}
            </div>
          </div>
        </div>
        <DashboardRenderer
          disableReportScrolling
          loadReport={evaluateReport}
          reports={reports}
          reportAddons={[
            <DragBehavior
              key="DragBehavior"
              reports={reports}
              updateReport={this.updateReport}
              onDragStart={this.hideAddButton}
              onDragEnd={this.showAddButton}
            />,
            <DeleteButton key="DeleteButton" deleteReport={this.deleteReport} />,
            <ResizeHandle
              key="ResizeHandle"
              reports={reports}
              updateReport={this.updateReport}
              onResizeStart={this.hideAddButton}
              onResizeEnd={this.showAddButton}
            />
          ]}
        >
          <Grid reports={reports} />
          <DimensionSetter emptyRows={9} reports={reports} />
          <AddButton addReport={this.addReport} visible={this.state.addButtonVisible} />
        </DashboardRenderer>
      </div>
    );
  }
}
