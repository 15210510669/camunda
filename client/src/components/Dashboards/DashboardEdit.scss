@import 'shared-styles/shared-styles.scss';

.DashboardEdit {
  .EntityNameForm {
    .separator {
      width: 1px;
      height: 23px;
      background-color: $silver-darken-87;
      margin-left: 4px;
      margin-right: 10px;
      transform: translate(0, 6px);
    }
  }

  .OptimizeReport .visualization,
  .ExternalReport iframe {
    opacity: 0.5;
  }
}
