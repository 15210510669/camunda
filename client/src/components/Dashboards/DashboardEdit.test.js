/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import {shallow} from 'enzyme';

import DashboardEdit from './DashboardEdit';

it('should contain a Grid', () => {
  const node = shallow(<DashboardEdit />);

  expect(node.find('Grid')).toBePresent();
});

it('should contain an AddButton', () => {
  const node = shallow(<DashboardEdit />);

  expect(node.find('AddButton')).toBePresent();
});

it('should editing report addons', () => {
  const node = shallow(<DashboardEdit />);

  expect(node.find('DashboardRenderer').prop('reportAddons')).toMatchSnapshot();
});

it('should hide the AddButton based on the state', () => {
  const node = shallow(<DashboardEdit />);
  node.setState({addButtonVisible: false});

  expect(node.find('AddButton').prop('visible')).toBe(false);
});

it('should select the name input field if dashboard is just created', () => {
  const node = shallow(<DashboardEdit isNew />);

  expect(node.find('EntityNameForm').prop('autofocus')).toBe(true);
});
