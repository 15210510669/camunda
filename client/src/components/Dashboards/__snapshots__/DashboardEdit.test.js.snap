// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should editing report addons 1`] = `
Array [
  <DragBehavior
    onDragEnd={[Function]}
    onDragStart={[Function]}
    updateReport={[Function]}
  />,
  <DeleteButton
    deleteReport={[Function]}
  />,
  <ResizeHandle
    onResizeEnd={[Function]}
    onResizeStart={[Function]}
    updateReport={[Function]}
  />,
]
`;
